src/CommandLine.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef COMMANDLINE_H
+#define COMMANDLINE_H
+
+#include "sbuf/SBuf.h"
+#include <vector>
+
+/// Manages arguments passed to the program, including the program name.
+/// The same info is passed to main() as argc and argv[] parameters.
+class CommandLine
+{
+public:
+    // codes for options without short option characters
+    enum LongCodes {
+        ForegroundCode = 1,
+        KidCode = 2
+    };
+
+    typedef std::pair<char, SBuf> OptionsPair;
+    typedef std::list<OptionsPair> Options;
```

Comment:
The list of options is needed for processing options. Unparsed argv_ is not sufficient for processing options because it does not have full 'option codes' list.

Reply:
I do not see how the `argv` parser's need to know "option codes" justifies having two types for options storage and/or storing two sets of options.

CommandLine user/caller will supply its "option codes" to the CommandLine when that user/caller wants to get parsed options (i.e., wants to access options by their names). CommandLine is a getopt() and `argv` wrapper, not a mainParseOptions() wrapper. In other words, this class should be CommandLine, not SquidExecutableCommandLine. We could create the latter class as well, but I do not think we should/need to do that.

If we make option effects/application CommandLine's business (as was done in the initial implementation), then even within the `main.cc` scope, there are already two different command lines: One for the master process and one for the kids. To start with, we need a general class that safely manipulates and parses `argv`. Its users will do their user-specific thing (adding more classes if needed, but I doubt it is needed).

Reply:
> I do not see how the argv parser's need to know "option codes"

Argv parser is currently getopt_long() method. This method obvously need "option codes", because it gets them as parameters (shortOpStr, squidOptions):
```
getopt_long(anArgc, anArgv, shortOpStr, squidOptions, &optIndex)
```

Reply:
> CommandLine is a getopt() and argv wrapper, not a mainParseOptions() wrapper.

It seems to me that only such 'single-purposed' CommandLine will not help much in this refactoring. For example, we need an effective way to configure/pass available options list to it. Currently, these are static string/array in mainParseOptions(), which are passed to getopt_long(). It would look awkwardly if we pass these getopt-specific data to CommandLine constructor. Therefore, we would need to add some getopt-independent options wrappers. For example, there should be an Option class (a single option configuration), ParsedOption (Option+parsed value), an OptionContainer (available options list), a OptionsConverter (to convert OptionContainer to getopt_long() -specific parameters), ParsedOptionsConverter (to convert parsed options back to argv list).
Are we going to do all these work?

Reply:
This is getting far too complicated. I think the solution(s) being proposed are loosing sight of the actual problem intending to be solved - which was apparently about Squid generating magic argv[0] values for execv() kid processes? There must be something missing because AFAIK Squid uses fork() for kid's and execv() only for helpers.

I would like some clarification please on exactly what the valgrind problem is. What code specifically is leading to the problem?

Reply:
> This is getting far too complicated.

Agreed. I am just having a hard time explaining how to remove all this unnecessary complexity. I will try to push some code changes instead.

> Squid uses fork() for kid's

It also uses execvp():

```
            if ((pid = fork()) == 0) {
                /* child */
                openlog(APP_SHORTNAME, LOG_PID | LOG_NDELAY | LOG_CONS, LOG_LOCAL4);
                prog = argv[0];
                argv[0] = const_cast<char*>(kid.name().termedBuf());
                execvp(prog, argv);
                int xerrno = errno;
                syslog(LOG_ALERT, "execvp failed: %s", xstrerr(xerrno));
            }
```

... which Valgrind wraps. Evidently, the Valgrind wrapper replaces our custom argv0 with the name of the executable, preventing Squid kids from getting their role and ID info. Please see the updated PR description. I think it explains it fairly well (but I am biased).

Reply:
> Alex: I do not see how X justifies Y.
> Eduard: X is obviously true.

I know. I was saying that X does not justify Y. I will try to push some code to move us closer to a simple solution. If you have any pending changes, please push them now.

Reply:
Has passing the role and ID through environment been considered? that would avoid all these issues altogether and allow debugging to run targeted kids individually as a bonus.

Reply:
> Has passing the role and ID through environment been considered?

Not until now AFAIK. Compared to a command line option, using environment is an inferior solution IMO because

* using environment introduces a new (or, to be more precise, promotes an obscure) configuration vector for routine SMP purposes;
* a running process environment is less visible to a sysadmin and most runtime debugging/logging tools;
* using environment gives no advantages compared to a command line option (that I can see).

> that would avoid all these issues altogether

While it would avoid the necessity to polish argv handling code, that polishing is a plus, not a negative. We should be happy that somebody will polish that code!

>  and allow debugging to run targeted kids individually as a bonus.

The proposed solution (i.e., the new --kid option) allows for that as well, using the usual command line configuration interface. I was actually quite excited about that! It is possible to manually start a given kid today as well, but faking argv0 is much harder than using a command line option, of course.

Reply:
> I know. I was saying that X does not justify Y. 

Evidently, it is clear that having two arrays for storing same data is not perfect. I was not going to argue with this obvious issue. Let's recollect that I suggested having only _parsed options array_ and generate the argv when required (I had almost implemented it), but after discussion, we rejected this approach (due to complexities with just-long options).

> I will try to push some code to move us closer to a simple solution. If you have any pending changes, please push them now.

Please push, I do not mind.